### PR TITLE
feat(builder): improve type of `BuildExecuter`

### PR DESCRIPTION
**Pull request by Asuka109**
# PR Details

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/builder-webpack-provider/src/shim.d.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+declare module 'webpack/lib/Stats' {
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/shim.d.ts:1]: 这里需要 shim 的原因是啥

File: packages/builder/builder-webpack-provider/src/core/build.ts [hunk 2]
```diff
@@ -9,33 +9,36 @@ import type {
...
       // When using run or watch, call close and wait for it to finish before calling run or watch again.
       // Concurrent compilations will corrupt the output files.
       compiler.close(closeErr => {
-        if (closeErr) {
-          logger.error(closeErr);
-        }
+        closeErr && logger.error(closeErr);
         if (err || !stats || stats.hasErrors()) {
-          const buildError: WebpackBuildError =
+          const buildError: WebpackBuildError<any> =
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/core/build.ts:35]: 可以没有 any 么~

Comment (chenjiahan) [reply packages/builder/builder-webpack-provider/src/core/build.ts:35]: 比如用 unknown 或者泛型？

File: packages/builder/builder-webpack-provider/src/shim.d.ts (hunk 1 above)
Comment (Asuka109) [reply packages/builder/builder-webpack-provider/src/shim.d.ts:1]: 补全 webpack 缺失的类型，其实和这个 pr 的主题无关，本来想用来写个 `isStats` 类型守卫后来感觉没太必要，这里留着方便以后用上 Stats

File: packages/builder/builder-webpack-provider/src/core/build.ts (hunk 2 above)
Comment (Asuka109) [reply packages/builder/builder-webpack-provider/src/core/build.ts:35]: 这里 any 对用户影响不大，catch 获得的 e 本身是无类型的，用户需要自己断言（这么说的话 `WebpackBuildError` 好像根本也不需要泛型

File: packages/builder/builder-webpack-provider/src/shim.d.ts (hunk 1 above)
Comment (chenjiahan) [reply packages/builder/builder-webpack-provider/src/shim.d.ts:1]: 不太明白为什么需要 declare module 'webpack/lib/Stats'...

File: packages/builder/builder-webpack-provider/src/core/build.ts (hunk 2 above)
Comment (chenjiahan) [reply packages/builder/builder-webpack-provider/src/core/build.ts:35]: 倒不是考虑对用户的影响，只是我们自己的代码里尽量避免出现 any，除非是没什么替代方式

File: packages/builder/builder-webpack-provider/src/shim.d.ts (hunk 1 above)
Comment (Asuka109) [reply packages/builder/builder-webpack-provider/src/shim.d.ts:1]: 看了眼原来 Stats 是有从主入口导出的... 那我把这段删了

File: packages/builder/builder-webpack-provider/src/core/build.ts (hunk 2 above)
Comment (Asuka109) [reply packages/builder/builder-webpack-provider/src/core/build.ts:35]: 想了下还是去掉泛型比较好 我改下哈

Comment (chenjiahan) [reply packages/builder/builder-webpack-provider/src/core/build.ts:35]: ok~

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**